src/cpp/exe/presolve/main.cpp

```diff
@@ -0,0 +1,179 @@
+    CouplingMap reduced_couplings = full_couplings;
+
+    const size_t nb_prob_total{full_couplings.size() - 1};
+    for (int nb_prob{0}; const auto& [filename, var_map]: full_couplings)
+    {
+        if (filename == options.MASTER_NAME) [[unlikely]]
+        {
+            // Keep master indices untouched
+            // and only try to reduce the subproblems
+            continue;
+        }
+
+        // Get all candidate indices in a sorted array
+        std::vector<int> colind(var_map.size());
+        std::transform(var_map.cbegin(),
+                       var_map.cend(),
+                       colind.begin(),
+                       [](const auto pair) { return pair.second; });
+        std::sort(colind.begin(), colind.end());
+
+        // Read full problem
+        const std::filesystem::path subproblem_path = input_root_dir / filename;
+
+        // TODO See about the keeprows thing
+        solver.read_prob_mps(subproblem_path);
+
+        // Keep the solver from removing candidate indices from subproblem
+        solver.mark_indices_to_keep_presolve(0, colind.size(), nullptr, colind.data());
+
+        solver.presolve_only();
+
+        // Move full subproblem to 'full' folder
+        // TODO Add option to keep or discard the full versions
+        // TODO All filesystem operations can be done in a separated part I guess
+        std::filesystem::rename(subproblem_path, full_dir / filename);
+
+        logger->display_message(
+          fmt::format("Subproblem '{}' reduced: {} / {}", filename, ++nb_prob, nb_prob_total));
+
+        // Write reduced problem MPS
+        solver.write_prob_mps(subproblem_path);
+
+        // Create a map [full_idx] -> reduced_idx for candidate indices
+        const auto full2reduced = get_presolve_map(solver, colind);
+
+        for (const auto& [var_name, idx]: var_map)
+        {
+            reduced_couplings.at(filename).at(var_name) = full2reduced.at(idx);
+        }
+    }
+
+    return reduced_couplings;
+}
+
+void write_reduced_problems(const CouplingMap& reduced_couplings,
+                            const BaseOptions& options,
+                            Logger logger)
+{
+    const auto structure_path{std::filesystem::path(options.INPUTROOT) / options.STRUCTURE_FILE};
+
+    // Write structure for reduced problem
+    export_structure_file(structure_path, reduced_couplings);
+    logger->display_message("Reduced " + options.STRUCTURE_FILE + " written");
+}
+
+int main(int argc, char** argv)
+{
+    usage(argc);
+    BaseOptions options{SimulationOptions(argv[1]).get_base_options()};
+    Logger logger = std::make_shared<xpansion::logger::User>(std::cout);
+
+    logger->display_message("Starting presolve");
+
+    XPRSPtr solver_ptr = init_solver(options, logger);
+
+    const CouplingMap reduced_couplings = reduce_problems(*solver_ptr, options, logger);
+
+    write_reduced_problems(reduced_couplings, options, logger);
```

Comment:
I do not understand why the writing part of the reduced structure is split from `reduce_problems` whereas the writing part for problems themselves is done within `reduce_problems`. Maybe do a consistant choice for all writing operations (especially if later on all this could be handled in-memory or not) ?

Reply:
To be honest, for no hard reason. Just to split into the different logical steps I saw ("solver initialization", "reduced coupling generation" and "output results"), but indeed it can go inside reduce_problems as the others :)